SerialFunctionDispatcher should implement isCurrent()

jyavenard opened the pull request:
#### cce771ba53901785917c014f762b1fe0a4d52df6
<pre>
SerialFunctionDispatcher should implement isCurrent()
<a href="https://bugs.webkit.org/show_bug.cgi?id=261008">https://bugs.webkit.org/show_bug.cgi?id=261008</a>
rdar://114801460

Reviewed by Kimmo Kinnunen.

- Expose isCurrent() to all existing SerialFunctionDispatcher and make
  assertIsCurrent() use isCurrent().
- Make use of assertIsCurrent() in places that were performing similar tests
  and streamline its use for consistency.

* Source/WTF/wtf/FunctionDispatcher.cpp:
(WTF::SerialFunctionDispatcher::assertIsCurrent const):
* Source/WTF/wtf/FunctionDispatcher.h:
* Source/WTF/wtf/RunLoop.cpp:
(WTF::RunLoop::assertIsCurrent const): Deleted.
* Source/WTF/wtf/RunLoop.h:
* Source/WTF/wtf/WorkQueue.cpp:
(WTF::WorkQueue::isCurrent const):
(WTF::WorkQueue::assertIsCurrent const): Deleted.
(WTF::WorkQueue::threadLikeAssertion const): Deleted.
* Source/WTF/wtf/WorkQueue.h:
* Source/WTF/wtf/cocoa/WorkQueueCocoa.cpp:
(WTF::WorkQueueBase::WorkQueueBase):
(WTF::WorkQueueBase::platformInitialize):
(WTF::WorkQueue::WorkQueue):
* Source/WTF/wtf/generic/WorkQueueGeneric.cpp:
(WTF::WorkQueueBase::WorkQueueBase):
(WTF::WorkQueueBase::platformInitialize):
* Source/WebCore/page/scrolling/ScrollingThread.cpp:
(WebCore::ScrollingThread::isCurrentThread):
* Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
(WebCore::AudioFileReader::handleMessage):
(WebCore::AudioFileReader::decodeAudioForBusCreation):
* Source/WebCore/platform/glib/FileMonitorGLib.cpp:
(WebCore::FileMonitor::FileMonitor):
(WebCore::FileMonitor::~FileMonitor):
* Source/WebCore/platform/graphics/ImageSource.cpp:
(WebCore::ImageSource::~ImageSource):
(WebCore::ImageSource::frameAtIndexCacheIfNeeded):
(WebCore::ImageSource::sourceSize):
* Source/WebCore/workers/WorkerOrWorkletThread.cpp:
(WebCore::WorkerOrWorkletThread::isCurrent const):
(WebCore::WorkerOrWorkletThread::assertIsCurrent const): Deleted.
* Source/WebCore/workers/WorkerOrWorkletThread.h:
* Source/WebKit/Platform/IPC/StreamConnectionWorkQueue.cpp:
(IPC::StreamConnectionWorkQueue::isCurrent const):
(IPC::StreamConnectionWorkQueue::assertIsCurrent const): Deleted.
* Source/WebKit/Platform/IPC/StreamConnectionWorkQueue.h:
(IPC::WTF_ASSERTS_ACQUIRED_CAPABILITY): Deleted.
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/CompositingRunLoop.cpp:
(WebKit::CompositingRunLoop::~CompositingRunLoop):
(WebKit::CompositingRunLoop::isCurrent const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Tools/TestWebKitAPI/Tests/WTF/WorkQueue.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/267808@main">https://commits.webkit.org/267808@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a49e8eaf1b2d36b5419c33148675fedd2c6d2021

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17737 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18067 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18604 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19561 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16590 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17932 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21358 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18217 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17950 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18249 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15413 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20425 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15482 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16164 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22731 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/15342 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16493 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16332 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20590 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/16949 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16898 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/14310 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/19307 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16012 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5204 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/4232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20376 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/20550 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16743 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5049 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
```diff
@@ -253,7 +253,7 @@ GstFlowReturn AudioFileReader::handleSample(GstAppSink* sink)
 void AudioFileReader::handleMessage(GstMessage* message)
 {
-    ASSERT(&m_runLoop == &RunLoop::current());
+    ASSERT(m_runLoop.isCurrent());
```
i think now it's better to use:
   assertIsCurrent(m_runLoop);
So you get the thread safety analysis should the future variable declarations start using it.

kkinnunen-apple commented on Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
```diff
@@ -389,7 +389,7 @@ void AudioFileReader::plugDeinterleave(GstPad* pad)
 void AudioFileReader::decodeAudioForBusCreation()
 {
-    ASSERT(&m_runLoop == &RunLoop::current());
+    ASSERT(m_runLoop.isCurrent());
```
i think now it's better to use:
   assertIsCurrent(m_runLoop);
So you get the thread safety analysis should the future variable declarations start using it.

kkinnunen-apple commented on Source/WebCore/platform/graphics/ImageSource.cpp:
```diff
@@ -64,7 +64,7 @@ ImageSource::ImageSource(Ref<NativeImage>&& nativeImage)
 {
     ASSERT(!hasAsyncDecodingQueue());
-    ASSERT(&m_runLoop == &RunLoop::current());
+    ASSERT(m_runLoop.isCurrent());
```
Alternatively this can store the runloop threadassertion as a member variable.

jyavenard replied:
I like `assertIsCurrent(m_runLoop);` better, no need to use `#if ASSERT_ENABLED`

kkinnunen-apple commented on Tools/TestWebKitAPI/Tests/WTF/WorkQueue.cpp:
```diff
@@ -285,16 +286,17 @@ TEST(WTF_WorkQueue, DestroyDispatchedOnDispatchQueue)
     private:
         std::atomic<size_t>& m_counter;
-        NO_UNIQUE_ADDRESS ThreadLikeAssertion m_ownerAssertion;
+        Ref<WorkQueue> m_queue;
```
I think this hunk you should remove, e.g. the functionality being used is the threadlikeassertion. So no need for the queue ref.

jyavenard replied:
the idea was for the test to work even with release build and we no longer need WorkQueue.threadLikeAssertion() be a public method

kkinnunen-apple commented on Tools/TestWebKitAPI/Tests/WTF/WorkQueue.cpp:
```diff
@@ -361,17 +363,13 @@ TEST(WTF_WorkQueue, ThreadSafetyAnalysisAssertIsCurrentWorks)
 TEST(WTF_WorkQueue, MAYBE_MainWorkQueueIsCurrent)
 {
-    assertIsCurrent(WorkQueue::main());
+    EXPECT_TRUE(WorkQueue::main().isCurrent());
```
Here too, the better and intended assertion is the `assertIsCurrent(WorkQueue::main());`.
So you get the thread safety analysis should the future variable declarations start using it.

kkinnunen-apple commented on Source/WTF/wtf/FunctionDispatcher.h:
```diff
@@ -45,8 +45,9 @@ class WTF_EXPORT_PRIVATE FunctionDispatcher {
+    virtual bool isCurrent() const = 0;
 #if ASSERT_ENABLED
-    virtual void assertIsCurrent() const = 0;
+    void assertIsCurrent() const;
```
Here now you can remove the assetIsCurrent method.

The intended contract is 
```
 assertIsCurrent([something that can be current]);
```

You'd then add the intended overload if none of the current overloads match.

jyavenard replied:
indeed. thanks

kkinnunen-apple reviewed (approved):
I'd prefer WorkQueue to not have isCurrent() on the count of that it enables pattern of:
```
  if (m_workQueue.isCurrent())
     doWork();
  else 
      m_workQueue.dispatch([] { doWork(); });
```

This is subtly incorrect in general case, as it runs the doWork() in the sync case before any other work in the queue. And queues are intended to preserve order.

However, if this makes your pending work very much easier, it may be justified and then we may need to look into this more if the risk realizes as bugs.
Use your own judgement based on this.

jyavenard committed:
SerialFunctionDispatcher should implement isCurrent()
https://bugs.webkit.org/show_bug.cgi?id=261008
rdar://114801460

Reviewed by Kimmo Kinnunen.

- Expose isCurrent() to all existing SerialFunctionDispatcher and make
  assertIsCurrent() use isCurrent().
- Make use of assertIsCurrent() in places that were performing similar tests
  and streamline its use for consistency.

* Source/WTF/wtf/FunctionDispatcher.cpp:
(WTF::SerialFunctionDispatcher::assertIsCurrent const):
* Source/WTF/wtf/FunctionDispatcher.h:
* Source/WTF/wtf/RunLoop.cpp:
(WTF::RunLoop::assertIsCurrent const): Deleted.
* Source/WTF/wtf/RunLoop.h:
* Source/WTF/wtf/WorkQueue.cpp:
(WTF::WorkQueue::isCurrent const):
(WTF::WorkQueue::assertIsCurrent const): Deleted.
(WTF::WorkQueue::threadLikeAssertion const): Deleted.
* Source/WTF/wtf/WorkQueue.h:
* Source/WTF/wtf/cocoa/WorkQueueCocoa.cpp:
(WTF::WorkQueueBase::WorkQueueBase):
(WTF::WorkQueueBase::platformInitialize):
(WTF::WorkQueue::WorkQueue):
* Source/WTF/wtf/generic/WorkQueueGeneric.cpp:
(WTF::WorkQueueBase::WorkQueueBase):
(WTF::WorkQueueBase::platformInitialize):
* Source/WebCore/page/scrolling/ScrollingThread.cpp:
(WebCore::ScrollingThread::isCurrentThread):
* Source/WebCore/platform/audio/gstreamer/AudioFileReaderGStreamer.cpp:
(WebCore::AudioFileReader::handleMessage):
(WebCore::AudioFileReader::decodeAudioForBusCreation):
* Source/WebCore/platform/glib/FileMonitorGLib.cpp:
(WebCore::FileMonitor::FileMonitor):
(WebCore::FileMonitor::~FileMonitor):
* Source/WebCore/platform/graphics/ImageSource.cpp:
(WebCore::ImageSource::~ImageSource):
(WebCore::ImageSource::frameAtIndexCacheIfNeeded):
(WebCore::ImageSource::sourceSize):
* Source/WebCore/workers/WorkerOrWorkletThread.cpp:
(WebCore::WorkerOrWorkletThread::isCurrent const):
(WebCore::WorkerOrWorkletThread::assertIsCurrent const): Deleted.
* Source/WebCore/workers/WorkerOrWorkletThread.h:
* Source/WebKit/Platform/IPC/StreamConnectionWorkQueue.cpp:
(IPC::StreamConnectionWorkQueue::isCurrent const):
(IPC::StreamConnectionWorkQueue::assertIsCurrent const): Deleted.
* Source/WebKit/Platform/IPC/StreamConnectionWorkQueue.h:
(IPC::WTF_ASSERTS_ACQUIRED_CAPABILITY): Deleted.
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/CompositingRunLoop.cpp:
(WebKit::CompositingRunLoop::~CompositingRunLoop):
(WebKit::CompositingRunLoop::isCurrent const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Tools/TestWebKitAPI/Tests/WTF/WorkQueue.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/267808@main
